aapcs64/aapcs64.rst

```diff
@@ -2305,6 +2328,188 @@ value of VG, with the subroutine having the following properties:
+
+**(Beta)**
+
+Platforms that support SME must provide a subroutine to save any state enabled
+by PSTATE.ZA.
+
+* The subroutine is called ``__arm_sme_save``.
+
+* The subroutine has a custom ``ZA`` `streaming-compatible interface`_ with
+  the following properties:
+
+  * X1-X15, X19-X29 and SP are call-preserved.
+  * Z0-Z31 are call-preserved.
+  * P0-P15 are call-preserved.
+
+* The subroutine takes the following arguments:
+
+  PTR
+    a 64-bit data pointer passed in X0 that points to a buffer which
+    is guaranteed to have a size that is equal to or larger than the size
+    returned by `__arm_sme_state_size`_. The pointer must be 16-byte aligned.
+
+* The subroutine does not return a value.
+
+* The subroutine behaves as follows:
+
+  * If ``PTR`` does not point to a valid buffer with the required size, the
+    behavior of calling this subroutine is undefined.
+
+  * If ZA state is 'active' on entry, then it is 'dormant' on normal return.
+    Otherwise the ZA state is unchanged.
+
+  * For the address ``PTR->VALID`` at an unspecified offset in the buffer,
+    the value 0 is written to ``PTR->VALID`` and the subroutine returns, if
+    either of the following conditions is true:
+
+    * The current thread does not have access to SME.
+
+    * PSTATE.ZA is 0.
+
+    * TPIDR2_EL0 is not a NULL pointer.
+
+  * For addresses ``PTR->BLK`` and ``PTR->ZA`` at unspecified offsets in
+    the buffer pointed to by ``PTR``, the routine must set up a lazy save
+    by zero-initializing the TPIDR2 block at address ``PTR->BLK``, then
+    writing address ``PTR->ZA`` to ``PTR->BLK.za_save_buffer``, writing the
+    streaming vector length in bytes (``SVL.B``) to
+    ``PTR->BLK.num_za_save_slices`` and finally copying the address ``PTR->BLK``
+    to ``TPIDR2_EL0``.
+
+  * If ZT0 is available, then for the address ``PTR->ZT0`` at an unspecified
+    offset in the buffer pointed to by ``PTR``, the contents of ZT0 are written
+    to ``PTR->ZT0``.
+
+  * The value 1 is written to ``PTR->VALID``.
+
+``__arm_sme_restore``
+^^^^^^^^^^^^^^^^^^^^^
+
+**(Beta)**
+
+Platforms that support SME must provide a subroutine to restore any state
+enabled by PSTATE.ZA.
+
+* The subroutine is called ``__arm_sme_restore``.
+
+* The subroutine has a custom ``ZA`` `streaming-compatible interface`_ with
+  the following properties:
+
+  * X1-X15, X19-X29 and SP are call-preserved.
+  * Z0-Z31 are call-preserved.
+  * P0-P15 are call-preserved.
+
+* The subroutine takes the following arguments:
+
+  PTR
+    a 64-bit data pointer passed in X0 that points to a buffer that
+    is initialized by a call to `__arm_sme_save`_. The pointer must be 16-byte
+    aligned.
+
+* The subroutine does not return a value.
+
+* The subroutine behaves as follows:
+
+  * If ``PTR`` does not point to a valid buffer with the required size, the
+    behavior of calling this routine is undefined.
+
+  * The ZA state on normal return is the same as the ZA state on entry to the
```

Comment:
This is no longer true if ZA was dormant or off, since the save might have been committed between the calls.  Perhaps we should just delete this bullet point and deal with the various cases below.  E.g. the next bullet point deals with the case where ZA was dormant or off on entry to the save function.